crystal i

asterite opened the pull request:
The experimental `crystal i` (interactive) command is here!

## What is `crystal i`

`crystal i` stands for "interactive crystal". It can be used in two modes right now:
1. `crystal i`: opens an interactive crystal session (REPL) similar to `irb` from Ruby
2. `crystal i file.cr`: runs a file in interpreted mode

In any of these two modes, you can use `debugger` in your code to debug it at that point. This is similar tu Ruby's [pry](https://github.com/pry/pry). There you can use these commands (similar to `pry` too):
- `step`: go to the next line/instruction, possibly going inside a method
- `next`: go to the next line/instruction, doesn't enter into methods
- `finish`: exit the current method
- `continue`: resume execution
- `whereami`: show where the debugger is

By the way, before this PR you could also use `debugger` and lldb or gdb would stop at that point, if compiled with `-d` (otherwise it seems the program just crashes, maybe we should just ignore `debugger` in that case: something to improve outside of this PR)

## Overview and some technical details

When running in interpreted mode, semantic analysis is done as usual, but instead of then using LLVM to generate code, we compile code to bytecode (custom bytecode defined in this PR, totally unrelated to LLVM). Then there's an interpreter that understands this bytecode. All of this for a stack-based machine. If you know nothing about interpreters and you want to learn more, I recommend reading the excellent https://www.craftinginterpreters.com/a-bytecode-virtual-machine.html 

The memory layout of things in interpreted mode matches the memory layout of compiled mode: a union of `Int32 | String` occupies 16 bytes in both modes, and there's an 8 bytes tag for the type ID. That said, the values for type IDs might not match between the two versions, and this makes the two worlds kind of incompatible (they can't interact with each other.)

C calls, and C callbacks, are implemented using [libffi](https://github.com/libffi/libffi) together with [dlopen](https://man7.org/linux/man-pages/man3/dlopen.3.html).

## Status

I'm just opening this PR to give a chance for others to start looking and commenting at the code, and ask questions if things are not clear. That said, I'd still like to take some time to clean up and comment a lot of the code before merging this. A lot of this was rushed because I wanted to have a big chunk ready for the conference.

Some things that are missing to implement related to the language:
- Raising and rescuing exceptions
- Closures
- Calling C functions that are outside of LibC (or outside the functions already included in the compiler, I _think_, because LibGC works)
- Some other primitives (some atomic functions, some math functions) and I'm sure some others more

And then some things that I know don't work fine right now:
- defining types and methods inside `crystal i` doesn't work super well, specially if you redefine methods or if you extend types (this is the real challenge, and we could maybe just postpone this one for how long as we want)
- it seems that defining variables when in `debugger` mode (let's call it `pry` from now on) doesn't work. I think this one should be easy to fix but I didn't have time to look into it
- free variables don't work well in `pry` don't work well (I'm not sure how to fix this, I think it requires a change also to the "real" Crystal)
- if an error happens (exception, or maybe something doesn't compile), I think it gets stuck in that mode (this should be easy to fix too)
- macro `finished` isn't executed yet
- `at_exit` isn't executed yet (because `main` isn't executed at all)

## Focus

For now, I'd like to focus on polishing the existing code and making sure it's 100% understandable and documented (some comments are already outdated, some method names are long and not nice, there might be some redundant code, and I'm sure there's already unclear code). After that we can merge this, put an "experimental" label next to the command, and then implement the missing things or fix existing bugs in separate PRs, which are going to be easier to review and merge like that.

I'm also open for suggestions on how to improve the code, be it for code clarity of for performance.

Also sorry beforehand for the commit (mis)organization and commit messages, this was all very experimental.

## Should this be in a separate repository?

This PR has some changes to the standard library, mainly some methods and functions need to be emulated in interpreted mode, and there's a new `flag?(:interpreted)` thing you can use. These could be in a separate PR, and then we could maybe extract the tool to a separate repository. But I think that having interpreted mode in the main repository, like every other tool, is fine from now. I believe moving tools outside of this repository is a separate discussion. But for instance, eventually `crystal play` could start using interpreted mode under the hood, or maybe you will be able to do `crystal spec -i ...` to run in interpreted mode. So, again, having all of this in this repository is probably fine.

## About FFI, dlopen and readline

~Right now I put them in the top-level just for simplicity. We should probably move them under the `Crystal` namespace because we don't want to publicly support these.~

There are now inside the `Crystal` namespace ✅ 

## More thoughts

- It's clear I never ran the full spec suit locally in this branch 😊 

## Steps before merging this PR (will be updated as we work on it)

- [ ] Install readline and libffi on CI
- [ ] Document every nook and cranny of the interpreter

asterite commented on src/compiler/crystal/command.cr:
```diff
@@ -64,7 +66,7 @@ class Crystal::Command
     when !command
       puts USAGE
       exit
-    when "init".starts_with?(command)
+    when command == "init"
```
This is technically a breaking change: if you did `crystal i` before it was the same as `crystal init`. Maybe it's fine, though? I don't know how many knew this.

bcardiff replied:
I think is fine, it been like that since the beginning but I never saw it used.

asterite commented on src/compiler/crystal/command/repl.cr:
```diff
@@ -0,0 +1,48 @@
+    OptionParser.parse(options) do |opts|
+      opts.banner = "Usage: crystal repl [options] [files]\n\nOptions:"
+
+      opts.on("-d", "--decompile", "Show decompilation") do
+        decompile = true
+      end
+
+      opts.on("-s", "--status", "Show time taken to execute") do
+        stats = true
+      end
+
+      opts.on("-t", "--trace", "Trace execution") do
+        trace = true
+      end
```
These are all experimental! Also `status` should be `stats` 😊 

I think the options there should match the options of the usual crystal compiled, being able to pass `--prelude`, `-D`, etc.

HertzDevil commented on src/compiler/crystal/types.cr:
```diff
@@ -719,6 +719,26 @@ module Crystal
       (yield self) ? true : false
     end
 
+    def always_falsey?
+      remove_indirection.is_a?(NilType)
+    end
+
+    def always_truthy?
+      type = remove_indirection
+      case type
+      when NilType
+        false
+      when BoolType
+        false
+      when PointerInstanceType
+        false
+      when UnionType
+        type.union_types.all? &.always_truthy?
+      else
+        true
+      end
+    end
+
```
I believe normal codegen would benefit from this part and it can be extracted into its own PR.

asterite replied:
Definitely! And this logic/optimization could be applied to While as well but it's currently missing from interpreted and compiled more.

asterite replied:
I think when I did this, I wanted to get something else working because I didn't have `is_a?` and `responds_to?` working. But it's still a good change :-)

bcardiff commented on src/readline/lib_readline.cr:
```diff
@@ -0,0 +1,19 @@
+{% if flag?(:openbsd) %}
+  @[Link("termcap")]
+{% end %}
+lib LibReadline
```
So this is reverting https://github.com/crystal-lang/crystal/pull/8364 as is. So there are a couple of places we might need to update the required libraries.

Maybe we should hide this inside `Crystal::` namespace, since we probably don't aim to do a full binding at least for now.

asterite replied:
Yes, after I sent the PR I updated the description, check the section "About FFI, dlopen and readline". I'll move these into the Crystal namespace 👍

asterite replied:
Done!

bcardiff commented on src/compiler/crystal/semantic/top_level_visitor.cr:
```diff
@@ -884,6 +880,8 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
     process_def_annotations(external, annotations) do |annotation_type, ann|
       if annotation_type == @program.call_convention_annotation
         call_convention = parse_call_convention(ann, call_convention)
+      elsif annotation_type == @program.primitive_annotation
```
Would it be better to keep the rule that the [Primitive] annotation should have no method body and write the code like:

```crystal
{% if flag?(:interpreted) %} 
  @[Primitive(:interpreter_raise_without_backtrace)] 
  def raise_without_backtrace(exception : Exception) : NoReturn
{% else %}
  def raise_without_backtrace(exception : Exception) : NoReturn
    # ...
  end
{% end %}
```

Or is there other motivation to loose this restriction?

asterite replied:
The main reasons were:
- It's easier to just say "This method is a primitive in interpreter mode"
- Without this change, the entire body is now defined inside a macro, and so if an error happens the error trace will expand the macro. Error traces inside macros don't look as nice as regular traces.
- Primitive annotations are only used by us, the std/compiler developers. Others don't need to worry about this. And I think lifting this restrictions makes our lives a bit easier. It would be pretty strange to have a primitive with a body, a `Primitive` annotation, and for that to be unintended (like in this case)

Do you still think we should go back to the original behavior? Right now I don't have a preference.

bcardiff replied:
Let's keep this as is, it's fine. The "Error traces inside macros don't look as nice as regular traces" sold it for me.
And it's true that PrimitiveAnnotation is only aimed for internal usage so it's not really changing something for users.

bcardiff commented on src/ffi/lib_ffi.cr:
```diff
@@ -0,0 +1,84 @@
+require "./enums"
+
+@[Link("ffi")]
+lib LibFFI
```
Similar to readline, we could hide it in Crystal::. But maybe there are some benefits on leaving this in the std-lib.

But most important this is adding a new dependency I'm not sure is always available.
Should we have an opt-out compiler flag that will disable the repl and remove the need of ffi and readline?

asterite replied:
Let's hide it in the Crystal namespace for now.

Should there be a `-D` flag names `--with-interpreter` or `--without-interpreter`? I'm not sure whether the interpreter should be included by default or not. libffi and readline are pretty ubiquitous, so it would be easy to include them in every Crystal distribution.

bcardiff replied:
I think it's better an opt-out `--without-interpreter`. If something breaks package maintainers can remove the feature and are aware of that.

If we do opt-in then users will need to report maintainers about the missing feature in order to use it.

bcardiff replied:
For a start the CI fails because `readline` is not present 🙈 

>    /usr/bin/ld: cannot find -lreadline (this usually means you need to install the development package for libreadline)

straight-shoota replied:
Compiler dependencies don't need to be included in Crystal distributions. We don't ship LLVM in our distribution packages.

The compiler binary should obviously have these libraries linked statically. But we shouldn't ship the libraries separately in the default distribution packages.

They can obviously be included in the build packages. Otherwise, if you want to build the compiler (with these features), you need to install the dependencies. Most Crystal users don't want to build the compiler.

So, there should be flags similar to `without_zlib` and `without_openssl` to disable the features that would link against these libraries (i.e. `crystal i`).

asterite replied:
How can I install readline on CI? Also feel free to push commits in the branch

Maybe I can move the branch to this repo so it's easier to contribute to?

bcardiff replied:
After 1.1 release we can rebuild those images to include manually readline on those and be sure next images will install it.

Sija commented on src/compiler/crystal/semantic/semantic_visitor.cr:
```diff
@@ -25,6 +25,24 @@ abstract class Crystal::SemanticVisitor < Crystal::Visitor
+    old_in_lib = @in_lib
+    old_in_c_struct_or_union = @in_c_struct_or_union
+    old_in_is_a = @in_is_a
+
+    yield
+
+    @current_type = old_current_type
+    @scope = old_scope
+    @exp_nest = old_exp_nest
+    @in_lib = old_in_lib
+    @in_c_struct_or_union = old_in_c_struct_or_union
+    @in_is_a = old_in_is_a
```
Shouldn't this be guarded with `ensure` block?

```suggestion
    begin
      yield
    ensure
      @current_type = old_current_type
      @scope = old_scope
      @exp_nest = old_exp_nest
      @in_lib = old_in_lib
      @in_c_struct_or_union = old_in_c_struct_or_union
      @in_is_a = old_in_is_a
```

asterite replied:
Oh, great point. I think that's why what I was trying to achieve wasn't working (this enables restoring the current type and scope information in case of an error, but I didn't spend too much time on this because back then it was a nice to have)

Sija commented on src/compiler/crystal/types.cr:
```diff
@@ -719,6 +719,26 @@ module Crystal
+    def always_truthy?
+      type = remove_indirection
+      case type
+      when NilType
+        false
+      when BoolType
+        false
+      when PointerInstanceType
+        false
```
Could be more succinct.

```suggestion
      when NilType, BoolType, PointerInstanceType
        false
```

asterite replied:
I'll consider it. Sometimes I like to see things listed line by line.

Sija commented on src/crystal/readline/readline.cr:
```diff
@@ -0,0 +1,136 @@
+    end
+  end
+
+  def done
```
```suggestion
  def done?
```

Sija commented on src/intrinsics.cr:
```diff
@@ -1,31 +1,57 @@
 # Intrinsics as exported by LLVM.
 # Use `Intrinsics` to have a unified API across LLVM versions.
 lib LibIntrinsics
+  {% if flag?(:interpreted) %} @[Primitive(:interpreter_intrinsics_debugtrap)] {% end %}
```
Since this looks like a pretty common idiom, perhaps it's worth adding a specific annotation, like an `InterpreterPrimitive` to avoid `{% if flag?(:interpreted) %}` boilerplate?

asterite replied:
It's a good point, but semantic analysis runs on interpreted code which means regular Crystal will have to know about this new annotation. The way it's done now, all regular primitives are primitives in the interpreter, but some regular methods need to be primitives in the interpreter too. I think it's fine.

jwoertink commented on src/crystal/readline/lib_readline.cr:
```diff
@@ -0,0 +1,21 @@
+module Crystal
+  @[Link("readline")]
```
Does this mean `readline` would be moved back in to stdlib? (https://github.com/crystal-lang/crystal-readline)

asterite replied:
It depends on whether we'll want to definitely use readline for this (I guess yes, because not using it is a bit annoying when typing.) But it would live inside the `Crystal` namespace, hidden from the std, just like the current `Markdown` lib.

wontruefree commented:
I think this REPL would work well as the default behavior for `crystal eval`.  Right now, if you run `crystal eval` with no other arguments, it waits for stdin, and that is weird behavior.  I thought opening a default `$EDITOR` like Lucky would be a better default behavior, but I think opening a REPL might be a better default behavior for this command.  Also, instead of creating a new command, adding the REPL to `eval` consolidates similar behavior.

wontruefree commented:
It might also make sense to have all code executed by `eval` to go through the interpreter and move the ability to compile and run static code to the `run` command.

This would mean any input either stdin or any code that is piped to `eval` would be interpreted and any code passed to `run` would be compiled and executed.  This would be a conceptual change to `eval` but we already have `build` or `run` for compiling and running static code files and I think we could fit a better api into that.  This would also be more predicable and consistent for someone learning the crystal CLI.

asterite commented:
I cleaned up the code and added a ton of comments to it. I think it's much more understandable now. It's a bit rough around declaring local variables, and declaring local variables for `pry` doesn't work yet, but it can be fixed later on.

# Benchmarks

I didn't take time to optimize things yet, but I also tried to do things in a way that's not super slow either.

Comparing it with Ruby... this code:

```ruby
# Ruby
time = Time.now
a = 0
while a < 10_000_000
  a += 1
end
puts Time.now - time
```

prints 0.111025 in Ruby 3.0.0

And this:

```crystal
# Crystal
puts Time.measure {
  a = 0
  while a < 10_000_000
    a += 1
  end
}
```

prints 00:00:00.341396724 in Crystal.

So it's roughly 3 times slower than Ruby in this case... but you have to understand that in Ruby the above is doing a method lookup and evaluating that, while in Crystal's case there's no method lookup, and it's just adding two values in the stack. It might be because the main VM loop is highly optimized in Ruby.

But then we have code like this:

```ruby
# Ruby
time = Time.now
a = []
100_000.times do |i|
  a << i
end
puts Time.now - time
```

prints 0.004612 in Ruby, but this:

```crystal
# Crystal
puts Time.measure {
  a = [] of Int32
  100_000.times do |i|
    a << i
  end
}
```

prints 00:00:00.059866375, which is more than 10 times slower! 😮  I think the main reason is that pushing an element to an Array is all interpreted in Crystal, while in Ruby just the method call is interpreted, but the actual push happens in C.

You can also see this with Hash operations:

```ruby
# Ruby
time = Time.now
h = {}
100_000.times do |i|
  h[i] = i
end
puts Time.now - time
```

prints 0.01986, and:

```crystal
# Crystal
puts Time.measure {
  h = {} of Int32 => Int32
  100_000.times do |i|
    h[i] = i
  end
}
```

prints 00:00:00.973007709 😮 , so almost 100x times slower. Again, I believe this is mainly because the code to insert into a Hash is all C in Ruby, but it's all interpreted in Crystal.

This is just to make sure everyone's expectations are in place regarding the speed of the current interpreter (in the conference I saw a lot of comments saying "How can it be so fast?", and I told you it's slower than Ruby 😉 .)

I'm not too worried about this speed difference. Maybe that speed is good enough for web development, where IO is usually the bottleneck. And it's a relatively good speed too for debugging things, maybe. Maybe for specs it could work too (it's rare for "normal" code to push thousands of elements into a Hash or an Array.) For Advent of Code or such contests, not so much.

But there's still room for optimization, I believe! And the bytecode could be eventually optimized.

lbguilherme commented:
I can't express in words how awesome all of this is. Thank you.

I have a curiosity: Why creating a new bytecode machine instead of using LLVM JIT? I'm not familiar with it, but I assume it could allow reusing most of the existing codegen code. Do you do something that can't be translated into LLVM or is it just much harder to use?

asterite commented:
@lbguilherme Yes, that was one option I considered when I started experimenting with this. I never tried it, here's why:
- we used LLVM JIT for all codegen specs initially, but back then it was choking with exceptions, personality, or I can't remember what was it. Maybe it's fixed now?
- if we generate LLVM code, do you know if we can stop at any point (like we can now with debugger) and have full control over the stack, know the type of local variables, etc.?

I also wanted to learn how a VM works, now it's less of a mystery to me (of course there are other more complex things like JITs, the Java VM is probably a beast in this, but for now I'm happy with what I learned.)

But, by all means, feel free to experiment with using LLVM for this. I'm sure it can lead to much faster execution of code!

jwoertink commented:
> I can't express in words how awesome all of this is. Thank you.

Agreed. Slowness can always be optimized later. The fact that this even works is already amazing. Great work on this @asterite ❤️

asterite commented:
To give an example, when you do this in the REPL:

```crystal
a = 1
```

there's a space in the stack for exactly 4 bytes for the int. But if you then do:

```crystal
a = rand < 0.5 ? a : "hello"
```

then `a` becomes a union, and the stack space for the local variables is adjusted a bit: the space for `a` now becomes 16 bytes to hold a union of Int32 or String.

Because I can control the stack, I can do this. I'm not sure how this can be done if we were to use LLVM JIT. Maybe you can pass a pointer to the stack memory?

wontruefree commented:
@asterite What do you think about these 2 comments about moving the command?

https://github.com/crystal-lang/crystal/pull/10910#issuecomment-879128346
https://github.com/crystal-lang/crystal/pull/10910#issuecomment-880106187

asterite commented:
I think using eval instead of i could work, I'm not sure. Changing that means it's a breaking change of some sort so I'd like to avoid that. Also i is shorter and more familiar, maybe.

bcardiff reviewed (commented):
Oopps I have a couple of comments from yesterday I didn't submit.

beta-ziliani commented:
About changing `eval`, I'd say that once we have this merged, we can simply deprecate it and point at the shinny new `i` command.

asterite commented on src/crystal/ffi/enums.cr:
```diff
@@ -0,0 +1,41 @@
+module Crystal::FFI
+  # TODO: define well based on architecture
```
If I remember correctly, the ABI values were different depending on the architecture. Maybe it's fine if we only use the DEFAULT value.

asterite commented on src/compiler/crystal/repl/syntax_highlighter.cr:
```diff
@@ -0,0 +1,189 @@
+require "./repl"
+require "colorize"
+
+module Crystal::Repl::SyntaxHighlighter
```
Not yet

Sija commented:
I'm not sure that's going to be a worthy change, since `eval` has its purpose for one liners.

jhass commented:
`crystal eval` is also useful for running code from standard input, `something_generating_crystal_code | crystal eval`.

asterite commented:
I started working on closures. The samples for `tcp_server.cr` and `tcp_client.cr` now work without modifications 🎉  (though they eventually raise, and that's not working correctly)

Also `channel_primes.cr` and `channel_select.cr` now work too.

asterite commented:
A few more commits in!

I worked on:
- more closure stuff (I think it's now all done, but I might be wrong)
- lib calls for libs other than libc
- exception handling (still missing doing things like `return` when there's `ensure`)
- exception backtraces, so `caller` now works, and exceptions that reach the top-level are shown nicely

Backtraces even work better than compiled mode, as you'll see once this is released.

Regarding libs, the SDL samples now work in interpreted mode, though super slow. Also the LLVM sample works too.

Still a lot is missing, but it's always less and less!

asterite commented:
Oh, and now `samples/tcp_server.cr` and `samples/tcp_client.cr` work exactly the same as compiled mode, thanks to exception handling.

j8r commented:
Is it planned to be able to easily embed the byte interpreter inside any regular Crystal program, with a public API? Say, like mruby.

asterite commented:
Not for the moment. My main goal with this is to be able to debug programs, and as a repl.

asterite commented:
A few more commits in!

Now [samples/http_server.cr](https://github.com/crystal-lang/crystal/blob/master/samples/http_server.cr) works in interpreted mode! (!!!)

From now on, to celebrate these little successes/achievements, I'll add a link to a youtube sound. Here's the first one for HTTP server (please turn down the volume before openning):

https://www.youtube.com/watch?v=eG-BOuN41vU

Out of curiosity I ran `ab` and `wrk` against the sample HTTP server in interpreted mode, and compared it against Ruby's sinatra (Ruby 2.6.6), just to get a feeling of how well it compares with something that's used in local development, to see if the slowness is going to be acceptable.

Here's the Ruby code:

```ruby
# Gemfile
source "http://rubygems.org"

gem 'rake'

gem 'sinatra', require: 'sinatra'
```

```ruby
# server.rb
# Run like this: ruby server.rb

# Requires the Gemfile
require 'bundler' ; Bundler.require

# By default Sinatra will return the string as the response.
get '/hello-world' do
  "Hello World!"
end
```

Ruby runs on port 4567, Crystal on 8080.

```
# This is Crystal, using wrk
$ wrk -t12 -c400 -d10s http://127.0.0.1:8080/
Running 10s test @ http://127.0.0.1:8080/
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   134.33ms  202.60ms   1.67s    91.18%
    Req/Sec   212.88    267.23     3.05k    94.92%
  22850 requests in 10.09s, 2.20MB read
  Socket errors: connect 155, read 209, write 0, timeout 0
Requests/sec:   2263.72
Transfer/sec:    223.28KB

# This is Ruby running wrk
$ wrk -t12 -c400 -d10s http://127.0.0.1:4567/hello-world
Running 10s test @ http://127.0.0.1:4567/hello-world
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    78.62ms   45.72ms 274.50ms   70.23%
    Req/Sec   116.01     53.81   272.00     63.98%
  12714 requests in 10.10s, 3.57MB read
  Socket errors: connect 155, read 217, write 0, timeout 0
Requests/sec:   1259.03
Transfer/sec:    361.70KB

# This is Crystal running ab
ab -k -c 100 -n 10000 127.0.0.1:8080/
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        127.0.0.1
Server Port:            8080

Document Path:          /
Document Length:        12 bytes

Concurrency Level:      100
Time taken for tests:   6.147 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    10000
Total transferred:      1010000 bytes
HTML transferred:       120000 bytes
Requests per second:    1626.76 [#/sec] (mean)
Time per request:       61.472 [ms] (mean)
Time per request:       0.615 [ms] (mean, across all concurrent requests)
Transfer rate:          160.45 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       4
Processing:     0   61  66.3     56     732
Waiting:        0   61  66.3     56     732
Total:          0   61  66.4     56     732

Percentage of the requests served within a certain time (ms)
  50%     56
  66%     62
  75%     71
  80%     79
  90%     95
  95%    133
  98%    195
  99%    385
 100%    732 (longest request)

# This is Ruby running ab
ab -k -c 100 -n 10000 127.0.0.1:4567/hello-world
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        WEBrick/1.4.2
Server Hostname:        127.0.0.1
Server Port:            4567

Document Path:          /hello-world
Document Length:        12 bytes

Concurrency Level:      100
Time taken for tests:   8.247 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    10000
Total transferred:      2940000 bytes
HTML transferred:       120000 bytes
Requests per second:    1212.60 [#/sec] (mean)
Time per request:       82.468 [ms] (mean)
Time per request:       0.825 [ms] (mean, across all concurrent requests)
Transfer rate:          348.15 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       3
Processing:     1   82  52.4     82     456
Waiting:        0   78  50.6     77     456
Total:          1   82  52.4     82     457

Percentage of the requests served within a certain time (ms)
  50%     82
  66%     98
  75%    115
  80%    128
  90%    149
  95%    173
  98%    198
  99%    218
 100%    457 (longest request)
```

So, yes, interpreted Crystal is slow... BUT! Right now it's faster than Ruby's sinatra, at least in version 2.6.6. So I think that's a good sign. And there's still room for optimizations!

crimson-knight commented:
@asterite am I reading the results wrong? It looks like Sinatra is faster to me. Honestly, Crystals slower performance is an acceptable speed considering this is for development and not production.

asterite commented:
@crimson-knight Yeah, I think the slow performance will be aceptable for development. That said, why do you think Ruby is faster than Crystal from the numbers above?

asterite commented:
@crimson-knight Oh, I see what you mean now. I was looking at "Requests per second" (Crystal has more than Ruby) and "Time per request" (Crystal takes less time than Ruby) but "Transfer rate" was higher for Ruby than Crystal! I checked with `curl -v` and it turns out Sinatra includes a bunch of default headers like "X-Frame-Options" that aren't included on the Crystal counterpart. So I added those headers to the Crystal server for a more fair comparison:

```crystal
# A very basic HTTP server
require "http/server"

server = HTTP::Server.new do |context|
  context.response.content_type = "text/html;charset=utf-8"
  context.response.headers["X-Xss-Protection"] = "1; mode=block"
  context.response.headers["X-Content-Type-Options"] = "nosniff"
  context.response.headers["X-Frame-Options"] = "SAMEORIGIN"
  context.response.headers["Server"] = "WEBrick/1.4.2 (Ruby/2.6.6/2020-03-31)"
  context.response.headers["Date"] = "Mon, 09 Aug 2021 22:47:08 GMT"
  context.response.print "Hello World!"
end

address = server.bind_tcp 8080
puts "Listening on http://#{address}"
server.listen
```

Now we have this for ab:

```
# Crystal
Requests per second:    1115.15 [#/sec] (mean)
Time per request:       89.674 [ms] (mean)
Time per request:       0.897 [ms] (mean, across all concurrent requests)
Transfer rate:          319.08 [Kbytes/sec] received

# Ruby
Requests per second:    1217.66 [#/sec] (mean)
Time per request:       82.125 [ms] (mean)
Time per request:       0.821 [ms] (mean, across all concurrent requests)
Transfer rate:          349.60 [Kbytes/sec] received
```

And for wrk:

```
# Crystal
Requests/sec:   1344.32
Transfer/sec:    384.65KB

# Ruby
Requests/sec:   1319.73
Transfer/sec:    379.07KB
```

So it seems Ruby+Sinatra is slightly faster here, but it doesn't matter. For development purposes having a bare-bones request take a millisecond is very good. With more complexity it might still take a lot more, but probably still good for development, especially if you can debug requests.

crimson-knight commented:
@asterite I was originally looking at the "Time Per Request" which appears to be better, this second comparison though looks like they come much closer.

And that's pretty good considering how light Sinatra is compared to Rails in dev (noticeably faster in my humble opinion)

So yeah this is totally acceptable performance. Honestly, just make it work to start. Ruby itself isn't going to get dramatically faster unless it's JITed or pre-compiled like the new Sorbet compiler--and even then it won't get nearly the performance boost that Crystal has going from this interpreted mode to compiled binary

asterite commented:
Another update, another video/music:

https://www.youtube.com/watch?v=69AyYUJUBTg&ab_channel=DekuSprout

Replace the text there with this:

> You found the `at_exit` handlers!
> Now you can run specs in interpreted mode

Caveat: not all specs run yet, some segfault, some don't pass! It's bugfixing time from now on (and a few missing things I need to implement)

But take a look at this:

```
# First run, no cache
$ time bin/crystal spec/std/levenshtein_spec.cr
..................

Finished in 878 microseconds
18 examples, 0 failures, 0 errors, 0 pending
bin/crystal spec/std/levenshtein_spec.cr  4.28s user 0.72s system 149% cpu 3.337 total

# Second run, some LLVM stuff is cached, that's why it's faster
$ time bin/crystal spec/std/levenshtein_spec.cr
..................

Finished in 1.16 milliseconds
18 examples, 0 failures, 0 errors, 0 pending
bin/crystal spec/std/levenshtein_spec.cr  1.69s user 0.33s system 128% cpu 1.573 total

# Using interpreted mode
$ time bin/crystal i spec/std/levenshtein_spec.cr
Using compiled compiler at .build/crystal
..................

Finished in 22.97 milliseconds
18 examples, 0 failures, 0 errors, 0 pending
bin/crystal i spec/std/levenshtein_spec.cr  0.81s user 0.16s system 127% cpu 0.758 total
```

This is just a small example, but you can see that running specs in interpreted mode takes 4 times less than running specs without the compiler having anything cached, and about twice as fast when things are cached. Caveat again: I don't know how will this behave with more complex code.

But let's try some CSV parsing:

```
$ time bin/crystal spec/std/csv/csv_parse_spec.cr
....................

Finished in 1.1 milliseconds
20 examples, 0 failures, 0 errors, 0 pending

bin/crystal spec/std/csv/csv_parse_spec.cr  4.34s user 0.59s system 173% cpu 2.839 total

$ time bin/crystal spec/std/csv/csv_parse_spec.cr
....................

Finished in 1.12 milliseconds
20 examples, 0 failures, 0 errors, 0 pending
bin/crystal spec/std/csv/csv_parse_spec.cr  1.72s user 0.33s system 124% cpu 1.645 total

$ time bin/crystal i spec/std/csv/csv_parse_spec.cr
....................

Finished in 21.37 milliseconds
20 examples, 0 failures, 0 errors, 0 pending
bin/crystal i spec/std/csv/csv_parse_spec.cr  0.86s user 0.17s system 129% cpu 0.797 total
```

Almost the same result! Looks promising!

And remember, we get an overall speedup, but we can also drop `debugger` anywhere to start debugging things 😸

sdogruyol commented:
Great achievement, hooray for the potty @asterite 🎉

asterite commented:
Well... a lot more is working now.

However, `Process.fork` seems to kind of work in interpreted mode but it randomly explodes (segfaults) and I don't understand why. Maybe some issue with `libffi` and fork, no idea.

For now I'll abandon trying to fix `fork`, though that means process execution (the backtick, Process.new, Process.run, etc.) won't work until we make `fork` work without randomly failing, but I have no idea where to start looking at.

Maybe someone else will be able to figure this out :-)

asterite commented:
(In fact I might make a total pause on this PR until someone figures this out, otherwise it's a bit pointless to continue. The other big thing missing, aside from bug fixes, is executing `ensure` blocks when doing `return` inside blocks, but I have an idea on how to do that. But I'll probably do it once the `fork` issue is solved)

straight-shoota commented:
I've tried running the interpreter on linux. With a few simple adjustments, all self-contained specs pass. But as soon as the prelude is included, there's an error. I can't even figure out what it the original error is. The interpreter registers an `EscapingException`. Trying to print that results in an invalid memory access.

I'd welcome any help for digging into that.

If someone else wants to try, this is the patch I used to make it run on linux: https://gist.github.com/straight-shoota/6484f5b1fa5a5d0c46a932c3f9f8909f

asterite commented:
> But as soon as the prelude is included, there's an error.

Can you try to pinpoint where that error happens? The easiest way would be to put a `debugger` line somewhere and do `next` until it crashes, then start again and try to do `step` where it crashed, etc.

straight-shoota commented:
Okay, so the issue seems to be with calling fstat. I jerry-rigged something together because the symbol `fstat` could not be found. I don't know what's going on there, but aparently fstat is just a wrapper to an internal function and for some reason it doesn't work. Apparantly my workaround didn't work either. The symbol can be found but it errors when being called. 🤷 

The first exception is raised when initializing `STDOUT` (because there's a call to `fstat`). But that doesn't break the program yet. I believe that exception causes initialization of `STDERR` and when that raises as well, the interpreter fails. Maybe because it recognizes that it's still in the process of rescuing an exception and suddenly, there's another exception from the same scope?

straight-shoota commented:
My suspicion was correct, the following code triggers the `EscapingException` as described:
```cr
@[Primitive(:interpreter_raise_without_backtrace)]
def raise(exception : Exception) : NoReturn
end

def foo
  raise Exception.new
end

begin
  foo
rescue exc
  foo
end
```

asterite commented:
I think the issue is that when an exception escapes then we print it using `ex.inspect_with_backtrace(STDOUT)`, and because STDOUT didn't finish initializing that probably breaks in a bad way. We could change it to call `ex.inspect_with_backtrace` instead, then print that string using the STDOUT from the interpreter, not the interpreted program. At least to understand better what's going on. I'll push a commit for this soon.

straight-shoota commented:
I don't think it's necessary tampering with exception printing. Knowing that that's the reason is enough. We should be able to investigate the actual error cause in isolation.

But on a more general perspective, we might want to make sure to make the kernel setup in general (not just for the interpreter) more resilient to issues with setting up the standard streams.

asterite commented:
@straight-shoota Done! Can you try again?

That said, I think we should figure out what's the problem with `fstat`. Maybe try with a small program in isolation, with `prelude=empty` (there's no option to make that happen, but you can comment "prelude" and do `require "empty"` there)

straight-shoota commented:
As I said in my previous comment, the last comment isn't necessary. It doesn't help me with debugging.

asterite commented:
@straight-shoota Maybe we could pair on this one day to see why `fstat` is not working. Maybe the bindings to `libffi` aren't correct for linux.

asterite commented:
@straight-shoota I think `__fxstat64` fails because it takes 3 arguments, not 2: https://github.com/lattera/glibc/blob/master/sysdeps/unix/sysv/linux/fxstat64.c#L32

You could also try and see if `fstat` works in Ruby for you: https://carc.in/#/r/bumw

straight-shoota commented:
Yeah, I've figured as much. But even with assumingly correct arguments, it still doesn't work.
It also reproduces with compiled Crystal, so this is unrelated to the interpreter.

The Ruby example fails for me as well. It appears that different versions of libc (I assume it's glibc because carc.in runs on Arch Linux) may have an `fstat` symbol or not.
However, I'm confused because when linking a compiled Crystal program there seems to be no issue with finding `fstat`.

straight-shoota commented:
I've updated the gist with a patch that now works for me on x86_64-linux-gnu 🚀 (https://gist.github.com/straight-shoota/6484f5b1fa5a5d0c46a932c3f9f8909f)

asterite commented:
@straight-shoota Great! I just pushed that diff minus the change to `out _` because I made that work.

Thinking about next steps, this is what comes to my mind:
- Make CI pass
- Allow distributing this (we probably need to ship libffi and libreadline)
- Maybe add a flag to completely disable the interpreter in case people want to distribute it without it
- Include a comment in the command's description that makes it very clear that this is still an experimental tool
- Merge this PR

After that, there's still a lot more to do but then it will be easier because the PRs will be smaller and more focused. In particular:
- Running `std_specs` still fail for the interpreter. On my side I get that out of the 247 std spec files we have, 31 are failing for various reasons: some things aren't implemented (Int128, casting between different tuple types inside a union, class variables for virtual types), some specs crash, probably because of bugs, and some specs work incorrectly, again probably because of bugs.
- The interpreter has a lot of TODOs that need to be done. They probably aren't hit because the code to trigger them should be uncommon, but it should still be done
- Used as a REPL, things like `require "http/client"` don't work out of the box because of how new types are added to the compiler. There's no precedent of adding new types to an existing program (they are all added once the entire program is known) so some work needs to be done there
- When you do `debugger` some things don't work. For example free variables aren't found, instance vars inside `initialize` can't be inspected, and sometimes some vars aren't found when they should.

asterite commented:
@straight-shoota Oh, maybe it's better if we make another branch so that any of us can push to it? Otherwise... how are you going to make changes to see if CI works? (if that's what you eventually intend to do)

straight-shoota commented:
Yeah, it might be a good idea to develop this feature in a branch on this repo. We can even make pull requests against that. I think we have still some work for getting this in a shippable state.
Usually it's a maintenance burdon to keep a long-running feature branch up with master. But `crystal i` is mostly new additions and applies only very few minor changes to existing code. So it shouldn't be hard to keep it around for some time, at least until we have some more stability.
I wouldn't keep it like that forever though (until it's "ready"). Just until we can be sure that the majority of the language is supported and we can exactly define what is still missing.

Fryguy commented:
FWIW, the default options on a PR allow maintainers to push into PR branches.  If you, when creating the original PR, hadn't changed this checkbox, then anyone with commit access on `crystal-lang:master` should have been able to push into `asterite:repl`.

![](https://user-images.githubusercontent.com/52120/131868836-3933cd0a-66ee-42af-a611-f03563a661e9.png)

asterite commented:
@Fryguy Thanks!

I still think it's better to have a branch in this repo. Then for example you can crate a PR that targets that's branch, and letting others push to it is just simpler.

asterite commented:
(also checking out the branch locally is simpler, no need to add remotes...)

Sija commented:
There's one drawback - it duplicates the CI jobs.

asterite commented:
@Sija How?

Sija commented:
@asterite They're being run on both, `push` and `pull_request` events.
